File: engine/core/tests/test_infill.cpp
// test_infill.cpp — native dLLM infilling (a capability AR models structurally lack), backend-free.
// Mirrors lab/tests/test_infill.py: fill a masked gap between a prefix and suffix under full
// bidirectional attention; both fixed sides are preserved verbatim and the gap is fully filled.
#include "cloze/events.hpp"
#include "cloze/generate.hpp"
#include "fake_adapter.hpp"

#include <cstdio>
#include <stdexcept>
#include <string>
#include <variant>
#include <vector>

using namespace cloze;

#define CHECK(cond) do { if (!(cond)) { std::fprintf(stderr, "CHECK failed: %s (line %d)\n", #cond, __LINE__); return 3; } } while (0)

namespace {
template <class T>
bool any_of(const std::vector<Event>& evs) {
    for (const Event& e : evs) if (std::holds_alternative<T>(e)) return true;
    return false;
}
}  // namespace

int main() {
    const int MASK = 15;  // FakeAdapter(vocab=16) mask id

    // Fills the gap, preserves both sides; reason "length", new_tokens == gap.
    {
        FakeAdapter fake(16, /*eos=*/-1, /*eos_at=*/-1);
        const std::vector<int> prefix = {1, 2, 3};
        const std::vector<int> suffix = {7, 8};
        const int gap = 5;
        GenerateConfig cfg{/*max_new=*/gap, /*steps=*/8, /*block_len=*/0, /*topk=*/-1};
        GenerateResult r = infill(fake, prefix, suffix, gap, cfg);

        const int lo = (int)prefix.size(), hi = lo + gap;
        CHECK((std::vector<int>(r.board.begin(), r.board.begin() + lo) == prefix));         // prefix verbatim
        CHECK((std::vector<int>(r.board.begin() + hi, r.board.end()) == suffix));           // suffix verbatim
        for (int i = lo; i < hi; ++i) CHECK(r.board[i] != MASK);                             // gap fully filled
        CHECK(r.reason == "length");
        CHECK(r.new_tokens == gap);
        CHECK((int)r.generated.size() == gap);
        CHECK(std::holds_alternative<GenStarted>(r.events.front()));
        CHECK(std::holds_alternative<GenFinished>(r.events.back()));
        CHECK(any_of<TokensCommitted>(r.events));
    }

    // One-sided context is allowed (prefix-only and suffix-only — suffix-only fills from pos 0).
    {
        FakeAdapter fake(16, -1, -1);
        GenerateConfig cfg{3, 6, 0, -1};
        GenerateResult pre = infill(fake, {5, 6}, {}, 3, cfg);
        GenerateResult suf = infill(fake, {}, {5, 6}, 3, cfg);
        CHECK(pre.new_tokens == 3 && suf.new_tokens == 3);
        for (int v : pre.generated) CHECK(v != MASK);
        for (int v : suf.generated) CHECK(v != MASK);
    }

    // Input validation.
    {
        FakeAdapter fake(16, -1, -1);
        GenerateConfig cfg{4, 2, 0, -1};
        bool threw = false;
        try { infill(fake, {1}, {2}, 0, cfg); } catch (const std::invalid_argument&) { threw = true; }
        CHECK(threw);
        threw = false;
        try { infill(fake, {}, {}, 4, cfg); } catch (const std::invalid_argument&) { threw = true; }
        CHECK(threw);
    }

    std::printf("test_infill: all assertions passed\n");
    return 0;
}

File: engine/lab/tests/test_infill.py
"""Infill / editing (DESIGN: native dLLM fill-in-the-middle) — fill a masked gap between a
prefix and a suffix under full bidirectional attention. Torch-free via FakeAdapter."""

import pytest

from cloze_lab.generate import GenerateConfig, infill
from cloze_lab.models.fake import FakeAdapter
from cloze_lab.scheduler.events import GenFinished, GenStarted, TokensCommitted, TokensRevised
from cloze_lab.scheduler.policies import RemaskLowConf


def test_fills_gap_and_preserves_both_sides() -> None:
    adapter = FakeAdapter(seed=7)
    prefix = adapter.encode("def add")
    suffix = adapter.encode("return x")
    gap = 5
    result = infill(adapter, prefix, suffix, gap, GenerateConfig(max_new=gap, steps=8))

    board = result.board.tolist()
    mask = adapter.config.mask_token_id
    lo, hi = len(prefix), len(prefix) + gap
    assert board[:lo] == prefix  # prefix verbatim
    assert board[hi:] == suffix  # suffix verbatim — the fixed right-context
    assert mask not in board[lo:hi]  # gap fully filled
    assert result.events[-1].reason == "length"
    assert result.events[-1].new_tokens == gap


def test_validates_inputs() -> None:
    adapter = FakeAdapter(seed=1)
    with pytest.raises(ValueError, match="gap must be"):
        infill(adapter, adapter.encode("a"), adapter.encode("b"), 0, GenerateConfig(max_new=1, steps=2))
    with pytest.raises(ValueError, match="prefix or a suffix"):
        infill(adapter, [], [], 4, GenerateConfig(max_new=4, steps=2))


def test_emits_event_stream() -> None:
    adapter = FakeAdapter(seed=3)
    result = infill(adapter, adapter.encode("x"), adapter.encode("y"), 4, GenerateConfig(max_new=4, steps=6))
    assert isinstance(result.events[0], GenStarted)
    assert isinstance(result.events[-1], GenFinished)
    assert any(isinstance(e, TokensCommitted) for e in result.events)


def test_one_sided_context_is_allowed() -> None:
    adapter = FakeAdapter(seed=2)
    only_prefix = infill(adapter, adapter.encode("hello"), [], 3, GenerateConfig(max_new=3, steps=6))
    only_suffix = infill(adapter, [], adapter.encode("world"), 3, GenerateConfig(max_new=3, steps=6))
    assert only_prefix.events[-1].new_tokens == 3
    assert only_suffix.events[-1].new_tokens == 3


def test_infill_supports_revisions() -> None:
    adapter = FakeAdapter(seed=7)
    result = infill(
        adapter, adapter.encode("a"), adapter.encode("b"), 4,
        GenerateConfig(max_new=4, steps=12), reviser=RemaskLowConf(tau_revise=1.0, max_revisions=1),
    )
    assert any(isinstance(e, TokensRevised) for e in result.events)
    assert result.events[-1].reason in ("length", "steps_exhausted")  # terminates either way
